HIP/hip-513.md

```diff
@@ -0,0 +1,340 @@
+when correcting errors during smart contract development.
+
+Enabling estimations like gas to be consumed.
+
+## Rationale
+
+[HIP-435](./hip-435.md) allows the creation of sidecar record files containing
+detailed information about transaction executions without bloating the main
+transaction record stream. This allows us to externalise verbose debug information
+about smart contract transactions that was previously expensive to download and
+process every time.
+
+## User stories
+
+Аs a mirror node operator, I want to have the option to download verbose contract operations-related data
+like contract state changes, contract actions, or contract bytecode.
+
+## Specification
+
+We will introduce a couple of sidecar record types, one containing the minimum
+needed data that allows a user to fully replay the transaction and derive the same
+set of results, and the other containing an entire operation-by-operation trace.
+
+### Add ContractStateChanges sidecar record type
+
+Add and populate a new sidecar record of type `ContractStateChanges`. It will
+have a single field - repeated `ContractStateChange`, which will be populated
+whenever a transaction causes contract storage states to change.
+This field will include the changes across all contract storage and not just the
+initial contract in the transaction.
+
+```
+message ContractStateChanges {
+  repeated ContractStateChange contract_state_changes = 1;
+}
+```
+
+For each slot read or written by the smart contract transaction
+a `ContactStateChange` record should be added to the `ContractStateChanges`.
+
+### Add ContractAction sidecar record type
+
+Add and populate a new sidecar record of type `ContractActions`. It will
+have a single field - repeated `ContractAction` with fine-grained information about transaction. 
+Which will improve debugging and auditing.
+
+```
+/**
+ * The type of action described by the action proto.
+ */
+enum ContractActionType {
+    /**
+     * default non-value.
+     */
+    NO_ACTION = 0;
+
+    /**
+     * Most CALL, CALLCODE, DELEGATECALL, and STATICCALL, and first action of ContractCall/ContractCallLocal to deployed
+     * contracts. This does not include calls to system or precompiled contracts.
+     */
+    CALL = 1;
+
+    /**
+     * CREATE, CREATE2, and first action of ContractCreate.
+     */
+
+    CREATE = 2;
+
+    /**
+     * like Call, but to precompiled contracts (0x1 to 0x9 as of Berlin)
+     */
+    PRECOMPILE = 3;
+
+    /**
+     * Call, but to system contract like HTS or ERC20 facades over Token accounts
+     */
+    SYSTEM = 4;
+}
+
+message ContractActions {
+    repeated ContractAction contract_actions = 1;
+}
+
+/**
+ * A finer grained action with a function result. Sometimes called "internal transactions." The function call itself
+ * will be the first action in a list, followed by sub-action in the order they were executed.
+ */
+message ContractAction {
```

Comment:
What will uniquely identify a contract action? The consensus timestamp plus what additional field(s)? Shouldn't there be a slot key to help identify or should we just use the order of the actions in the list as an index so the primary key is `(consensus_timestamp, index)`?

Reply:
Ordinal nested order is the industry standard - https://openethereum.github.io/JSONRPC-trace-module#traceaddress-field.

We can reconstruct this index with the order of the contract actions and the `call_depth` attribute.  We can also simply do order of call within the transaction regardless of depth.  Array == nested ID integer = overall order.  We would also need the parent transaction ID.